extensions/amp-animation/0.1/css-expr-ast.js

```diff
@@ -0,0 +1,250 @@
+  constructor(num) {
+    super();
+    /** @const @private {number} */
+    this.num_ = num;
+  }
+}
+
+
+/**
+ * A CSS length value: `100px`, `80vw`, etc.
+ */
+export class CssLengthNode extends CssNode {
+  /**
+   * @param {number} num
+   * @param {string} units
+   */
+  constructor(num, units) {
+    super();
+    /** @const @private {number} */
+    this.num_ = num;
+    /** @const @private {string} */
+    this.units_ = units;
+  }
+}
+
+
+/**
+ * A CSS angle value: `45deg`, `0.5rad`, etc.
+ */
+export class CssAngleNode extends CssNode {
+  /**
+   * @param {number} num
+   * @param {string} units
+   */
+  constructor(num, units) {
+    super();
+    /** @const @private {number} */
+    this.num_ = num;
+    /** @const @private {string} */
+    this.units_ = units;
+  }
+}
+
+
+/**
+ * A CSS time value: `1s`, `600ms`.
+ */
+export class CssTimeNode extends CssNode {
+  /**
+   * @param {number} num
+   * @param {string} units
+   */
+  constructor(num, units) {
+    super();
+    /** @const @private {number} */
+    this.num_ = num;
+    /** @const @private {string} */
+    this.units_ = units;
+  }
+}
+
+
+/**
+ * A CSS generic function: `rgb(1, 1, 1)`, `translateX(300px)`, etc.
+ */
+export class CssFuncNode extends CssNode {
+  /**
+   * @param {string} name
+   * @param {!Array<!CssNode>} args
+   */
+  constructor(name, args) {
+    super();
+    /** @const @private {string} */
+    this.name_ = endsWith(name, '(') ?
+        name.substring(0, name.length - 1) : name;
+    /** @const @private {!Array<!CssNode>} */
+    this.args_ = args;
+  }
+}
+
+
+/**
+ * A CSS `var()` expression: `var(--name)`, `var(--name, 100px)`, etc.
+ * See https://www.w3.org/TR/css-variables/.
+ */
+export class CssVarNode extends CssNode {
+  /**
+   * @param {string} varName
+   * @param {!CssNode=} opt_def
+   */
+  constructor(varName, opt_def) {
+    super();
+    /** @const @private {string} */
+    this.varName_ = varName;
+    /** @const @private {?CssNode} */
+    this.def_ = opt_def || null;
+  }
+}
+
+
+/**
+ * A CSS `calc()` expression: `calc(100px)`, `calc(80vw - 30em)`, etc.
+ * See https://drafts.csswg.org/css-values-3/#calc-notation.
+ */
+export class CssCalcNode extends CssNode {
+  /** @param {!CssNode} expr */
+  constructor(expr) {
+    super();
+    /** @const @private {!CssNode} */
+    this.expr_ = expr;
+  }
+}
+
+
+/**
+ * A CSS `calc()` sum expression: `100px + 20em`, `80vw - 30em`, etc.
+ */
+export class CssCalcSumNode extends CssNode {
+  /**
+   * @param {!CssNode} left
+   * @param {!CssNode} right
+   * @param {string} op Either "+" or "-".
+   */
+  constructor(left, right, op) {
+    super();
+    /** @const @private {!CssNode} */
+    this.left_ = left;
+    /** @const @private {!CssNode} */
+    this.right_ = right;
+    /** @const @private {string} */
+    this.op_ = op;
+  }
+}
+
+
+/**
+ * A CSS `calc()` product expression: `100px * 2`, `80vw / 2`, etc.
+ */
+export class CssCalcProductNode extends CssNode {
```

Comment:
I could and may still do so later. But implementation-wise per spec they are super different. So not sure if I'll win anything if I do at this time.